docs/configuration-files.md

```diff
@@ -224,6 +226,64 @@ rules:
     severity: ERROR
 ```
 
+### `metavariable-comparison`
+
+The `metavariable-comparison` operator compares metavariables against a basic [Python comparison](https://docs.python.org/3/reference/expressions.html#comparisons)
+expression. This is useful for filtering results based on a [metavariable's](/docs/pattern-features.md#metavariables) value.
+
+**Example**
+
+The `metavariable-comparison` operator is a mapping which requires the
+`metavariable` and `comparison` keys. It can be combined with other pattern operators:
+
+```yaml
+rules:
+  - id: superuser-port
+    patterns:
+      - pattern: set_port($ARG)
+      - metavariable-comparison:
+          metavariable: '$ARG'
+          comparison: '$ARG < 1024'
+    message: "module setting superuser port"
+    languages: [python]
+    severity: ERROR
+```
+
+This will catch code like `set_port(80)` or `set_port(443)`, but not `set_port(8080)`.
+
+The `metavariable-comparison` operator also takes optional `base: int` and
```

Comment:
the number syntax itself contains the base, why specify the base here?
OCaml, and I guess Python will be able to read 0xa or 10 or 0o82 or whatever syntax and convert to an internal number.

Reply:
I don't think that's true at the moment. The regex for integer detection is currently `^[0-9]+$`, which misses `0x`, `0o`, etc. Since the metavariable content is always initially a string, I wanted to provide a mechanism for parsing the string value in different bases.

Generally, if you look at our `semgrep-rules`, the majority of rules using `pattern-where-python` perform a simple comparison, and often parse a number in a different base, or perform a string strip. I know this implementation isn't the maximally powerful/generalized solution, it instead aims to cover the majority of use cases, not all of them, and hit the sweet spot between power and usability (and minimize `--dangerously...` usage). E.g.

```
24:  - pattern-where-python: |
25-      int(vars['$PERM'], 8) > 0o600
```

```
18:  - pattern-where-python: |-
19-      int(vars['$BITS']) < 2048
```

```
18:  - pattern-where-python: |-
19-      int(vars['$SIZE']) < 128
```

```
11:  - pattern-where-python: int(vars['$NUM'].replace('"', '')) > 32767 or int(vars['$NUM'].replace('"', '')) < -32768
```

```
21:  - pattern-where-python: |-
22-      int(vars['$NUM'].replace('"', '')) > 2147483647 or int(vars['$NUM'].replace('"', '')) < -2147483648
```

Reply:
@aryx I had a thought here: would it make more sense for `semgrep-core` to simply take in the data as the corresponding JSON type instead of performing string parsing? I.e. a JSON integer is an integer, a JSON float is a float, a JSON string is a string, etc, instead of trying to interpret the data with regexs. That way Python can interpret the metavariable data it receives, then `semgrep-core` can simply operate on the JSON types.

IMO Python is better equipped to deal with things like string parsing, while OCaml is obviously better equipped to deal with program analysis tasks.

Reply:
Yes, good idea (but no Python is not better equipped for string parsing :) )

Reply:
I'll modify -eval to accept directly the right Json type.

Reply:
@mschwager https://github.com/returntocorp/semgrep/pull/1839